Translation for "Sharing State Between Components" (#872)
reactjs/ru.react.dev: src/content/learn/sharing-state-between-components.md

---
title: Совместное использование состояния между компонентами
---

<Intro>

Иногда требуется, чтобы состояние двух компонентов всегда изменялось вместе. Для этого нужно удалить состояние из обоих компонентов, переместить его в ближайшего общего родителя и передать его им через пропсы. Это называется *подъемом состояния вверх* и является одним из наиболее распространенных приемов при написании кода на React.

</Intro>

> <YouWillLearn>

Author: YouWillLearn, где-то отдельно будет переведен, правильно понимаю?
![Screenshot 2023-05-06 at 18 42 57](https://user-images.githubusercontent.com/15185407/236649273-751e43b1-a267-4f5a-a810-ac3f317c75cc.jpeg)

Reviewer: если есть возможность, можете поправить?

</Sandpack>

Обратите внимание, что нажатие кнопки на одной панели не влияет на другую панель--они независимы.

<DiagramGroup>

> 

Reviewer: пожалуйста удалить лишний перенос строку, чтобы в будущем не было проблем при ребейзе.

Чтобы согласовать поведение этих двух панелей, вам нужно "поднять их состояние" в родительский компонент в три шага:

1. **Удалите** состояние из дочерних компонентов.
2. **Передайте** данные хардкодом из общего родителя.
3. **Добавьте** состояние в общего родителя и передайте его вместе с обработчиками событий.

Это позволит компоненту `Accordion` управлять обеими панелями и раскрывать только по одной за раз.

### Шаг 1: Удалить состояние из дочерних компонентов {/*step-1-remove-state-from-the-child-components*/}

Вы передадите управление значением `isActive` родительскому компоненту `Panel`. Это означает, что родительский компонент будет передавать значение `isActive` через проп, вместо того, чтобы хранить это состояние в `Panel`. Начните с **удаления этой строки** из компонента `Panel`:

```js
const [isActive, setIsActive] = useState(false);
```

И вместо нее, добавьте `isActive` в список пропсов `Panel`:

```js
function Panel({ title, children, isActive }) {
```

> Теперь родитель компонента `Panel`  может *управлять* `isActive` [передавая его вниз как проп.](/learn/passing-props-to-a-component) С другой стороны, компонент `Panel` теперь *не имеет контроля* над значением `isActive`--теперь оно зависит от родительского компонента!

Reviewer: ```suggestion
Теперь родитель компонента `Panel` может *управлять* `isActive`, [передавая его вниз как проп.](/learn/passing-props-to-a-component) С другой стороны, компонент `Panel` теперь *не имеет контроля* над значением `isActive`--теперь оно зависит от родительского компонента!
```

</Sandpack>

Попробуйте изменить хардкод значение `isActive` в компоненте `Accordion` и обратите внимание как это повлияет на результат.

### Шаг 3: Добавить состояние в общего родителя {/*step-3-add-state-to-the-common-parent*/}

Подъём состояния часто приводит к изменению сущности хранимого состояния.

> В нашем случае, только одна панель должна быть активна одновременно. Это означает, что общему родителю `Accordion` нужно следить за тем *какая* панель активна в данный момент. Вместо булевого значения в состоянии можно хранить число, которое будет означать индекс активной панели:

Reviewer 2: ```suggestion
В нашем случае только одна панель должна быть активна одновременно. Это означает, что общему родителю `Accordion` нужно следить за тем *какая* панель активна в данный момент. Вместо булевого значения в состоянии можно хранить число, которое будет означать индекс активной панели:
```
Здесь также не требуется обособление

```js
const [activeIndex, setActiveIndex] = useState(0);
```

Когда `activeIndex` равен `0`, активна первая панель, а когда равен `1`, активна вторая.

> Нажатие на кнопку «Показать» в любой из панелей должно изменять индекс активной панели в `Accordion`. `Panel` не может напрямую установить состояние `activeIndex`, потому что оно определено внутри `Accordion`. Компоненту `Accordion` необходимо *явно разрешить* компоненту `Panel` изменять свое состояние, [передав обработчик события как проп](/learn/responding-to-events#passing-event-handlers-as-props):

Reviewer 2: ```suggestion
Нажатие на кнопку «Показать» в любой из панелей должно изменять индекс активной панели в `Accordion`. `Panel` не может напрямую установить состояние `activeIndex`, потому что оно определено внутри `Accordion`. Компонент `Accordion` должен *явно разрешить* компоненту `Panel` изменять свое состояние, [передав обработчик события как проп](/learn/responding-to-events#passing-event-handlers-as-props):
```
мне кажется, так попроще звучит

</Sandpack>

Подъём состояния завершен! Переместив состояние в общий родительский компонент нам удалось скоординировать две панели. Использование индекса активной панели вместо двух флагов `isActive` гарантирует нам, что будет активна только одна панель одновременно. А передав обработчики событий дочерним компонентам мы позволили им управлять состоянием родителя.

<DiagramGroup>

<Diagram name="sharing_state_parent" height={385} width={487} alt="На диаграмме изображено дерево компонентов, один родительский компонент с названием Accordion и два дочерних с названием Panel. Accordion содержит переменную activeIndex со значением ноль, которая преобразуется в isActive со значением true для первой Panel, и в isActive со значением false для второй Panel." >

> В начале, `activeIndex` в компоненте `Accordion` имеет значение 0, поэтому первая панель получает `isActive = true`

Reviewer 2: ```suggestion
Сначала `activeIndex` в компоненте `Accordion` имеет значение 0, поэтому первая панель получает `isActive = true`
```

</Diagram>

<Diagram name="sharing_state_parent_clicked" height={385} width={521} alt="Та же диаграмма, что и предыдущая, с выделенным значением активного индекса родительского компонента Accordion, указывающим на клик со значением, измененным на единицу. Стрелки к обоим дочерним компонентам Panel также выделены, а значение isActive, передаваемое каждому дочернему элементу, установлено в противоположное значение: false для первой панели и true для второй." >

Когда состояние `activeIndex` в `Accordion` становится `1`, вторая `Panel` получает `isActive = true` вместо первой

</Diagram>

</DiagramGroup>

<DeepDive>

#### Управляемые и неуправляемые компоненты {/*controlled-and-uncontrolled-components*/}

Компоненты, у которых есть внутреннее состояние, обычно называют «неуправляемыми». Например, оригинальный компонент `Panel` с переменной состояния `isActive` является неуправляемым, потому что его родитель не может повлиять на то, активна ли панель.

> Напротив, компонент называют «управляемым», если важная информация зависит от пропов, а не от внутреннего состояния. Это позволяет родительскому компоненту полностью контролировать его поведение. После наших изменений, компонентом `Panel` управляет компонент `Accordion` с помощью пропа `isActive`.

Reviewer 2: ```suggestion
Напротив, компонент называют «управляемым», если важная информация зависит от пропсов, а не от внутреннего состояния. Это позволяет родительскому компоненту полностью контролировать его поведение. После наших изменений, компонентом `Panel` управляет компонент `Accordion` с помощью пропа `isActive`.
```
по соглашению во множественном числе будет "пропсов"

Неуправляемые компоненты проще использовать, потому что им необходимо меньше конфигурации. Но они не такие гибкие, когда вам нужно согласовать их поведение. Управляемые компоненты максимально гибкие, но необходимо чтобы родительские компоненты полностью настраивали их через пропсы.

> На практике «управляемые» и «неуправляемые» это не строгие технические определения--обычно компоненты имеют и внутреннее состояние и пропсы. Однако, важно понимать как устроены компоненты и какие возможности они предоставляют.

Reviewer 2: ```suggestion
На практике «управляемые» и «неуправляемые» это не строгие технические определения--обычно компоненты имеют и внутреннее состояние, и пропсы. Однако, важно понимать как устроены компоненты и какие возможности они предоставляют.
```

Ваше приложение будет меняться по мере работы над ним. Это нормально, что вы будете перемещать состояние вниз или обратно вверх, пока разбираетесь где «живет» каждая часть состояния. Это все часть процесса!

Чтобы увидеть как это работает на практике с другими компонентами, прочитайте статью [Мыслим на React.](/learn/thinking-in-react).

<Recap>

* Если вы хотите скоординировать два компонента, переместите их состояние к общему родительскому компоненту.
* Затем передайте информацию через пропсы из общего родительского компонента.
* Наконец, пробросьте обработчики событий, чтобы дочерние компоненты могли изменять состояние родительского компонента.

> * Полезно рассматривать компоненты как «управляемые» (управляемые пропами) или «неуправляемые» (управляемые состоянием).

Reviewer 2: ```suggestion
* Полезно рассматривать компоненты как «управляемые» (управляемые пропсами) или «неуправляемые» (управляемые состоянием).
```
По соглашению пишем "пропсами"

Когда состояние `activeIndex` в `Accordion` становится `1`, вторая `Panel` получает `isActive = true` вместо первой

</Diagram>

</DiagramGroup>

<DeepDive>

#### Управляемые и неуправляемые компоненты {/*controlled-and-uncontrolled-components*/}

Компоненты, у которых есть внутреннее состояние, обычно называют "неуправляемыми". Например, оригинальный компонент `Panel` с переменной состояния `isActive` является неуправляемым, потому что его родитель не может повлиять на то, активна ли панель.

Напротив, компонент называют "управляемым", если важная информация зависит от пропов, а не от внутреннего состояния. Это позволяет родительскому компоненту полностью контролировать его поведение. После наших изменений, компонентом `Panel` управляет компонент `Accordion` с помощью пропа `isActive`.

Неуправляемые компоненты проще использовать, потому что им необходимо меньше конфигурации. Но они не такие гибкие, когда вам нужно согласовать их поведение. Управляемые компоненты максимально гибкие, но необходимо чтобы родительские компоненты полностью настраивали их через пропсы.

> На практике, "управляемые" и "неуправляемые" это не строгие технические определения--обычно компоненты имеют и внутреннее состояние и пропсы. Однако, важно понимать как устроены компоненты и какие возможности они предоставляют.

Reviewer: ```suggestion
На практике «управляемые» и «неуправляемые» это не строгие технические определения--обычно компоненты имеют и внутреннее состояние и пропсы. Однако, важно понимать как устроены компоненты и какие возможности они предоставляют.
```

Когда пишите компонент, подумайте какая информация в нем должна быть управляема (через пропсы), а какая неуправляема (через состояние). Но вы в любой момент можете передумать и сделать рефакторинг позже.

</DeepDive>

## Единый источник истины для каждого состояния {/*a-single-source-of-truth-for-each-state*/}

> В приложении на React у многих компонентов будет свое собственное состояние. Какое-то состояние может "жить" близко к листовым компонентам (компоненты внизу дерева), например, текстовое поле. Другое состояние может "жить" ближе к корню приложения. Например, даже при реализации библиотеки клиентской маршрутизации, текущий путь обычно сохраняют в состояние React и передают его вниз через пропсы.

Reviewer: ```suggestion
В приложении на React у многих компонентов будет свое собственное состояние. Какое-то состояние может «жить» близко к листовым компонентам (компоненты внизу дерева), например, текстовое поле. Другое состояние может «жить» ближе к корню приложения. Например, даже при реализации библиотеки клиентской маршрутизации, текущий путь обычно сохраняют в состояние React и передают его вниз через пропсы.
```

пожалуйста поменяйте тут и в других местах кавычки на «ёлочки»

> **Для каждого уникального кусочка состояния, вы будете выбирать, какой компонент будет им "владеть".** Этот принцип также известен как наличие ["единого источника истины".](https://en.wikipedia.org/wiki/Single_source_of_truth) Это не означает, что все состояние React находится в одном месте--для _каждой_ части состояния существует _конкретный_ компонент, который хранит эту информацию. 

Reviewer: ```suggestion
**Для каждого уникального кусочка состояния вы выбираете компонент, который «владеет» им** Этот принцип также известен как наличие ["единого источника истины".](https://en.wikipedia.org/wiki/Single_source_of_truth) Это не означает, что все состояние React находится в одном месте--для _каждой_ части состояния существует _конкретный_ компонент, который хранит эту информацию. 
```

  
Пожалуйста, проверьте запятые. Есть разница в правилах с английским языком.

> Вместо дублирования общего состояния между компонентами, *поднимите* его в общий родительский компонент и "пробросьте" его дочерним компонентам, которым он необходим.

Reviewer: Пожалуйста удалите перенос. Для корректного ребейза в будущем нужно чтобы строчки совпадали

Reviewer: ```suggestion
Вместо дублирования общего состояния между компонентами, *поднимите* его в общий родительский компонент и «пробросьте» его дочерним компонентам, которым он необходим.
```

> Ваше приложение будет меняться по мере работы над ним. Это нормально, что вы будете перемещать состояние вниз или обратно вверх, пока разбираетесь где "живет" каждая часть состояния. Это все часть процесса!

Reviewer: ```suggestion
Ваше приложение будет меняться по мере работы над ним. Это нормально, что вы будете перемещать состояние вниз или обратно вверх, пока разбираетесь где «живет» каждая часть состояния. Это все часть процесса!
```

> Чтобы увидеть, как это работает на практике с другими компонентами, прочитайте статью [Мышление в React.](/learn/thinking-in-react).

Reviewer: ```suggestion
Чтобы увидеть как это работает на практике с другими компонентами, прочитайте статью [Мыслим на React.](/learn/thinking-in-react).
```

Reviewer: Финальный заголовок этой страницы будет тут:
https://github.com/reactjs/ru.react.dev/pull/876

  return (
    <label>
      Поиск:{' '}
      <input
        value={query}

>         onChange={handleChange}

Reviewer 2: Вы могли бы перевести этот пример (массив `foods` в частности) полностью?
